File: frontend/apps/android/app/src/main/java/app/pantopus/android/ui/screens/shared/content_detail/headers/ProfileHeader.kt
@file:Suppress("MagicNumber", "PackageNaming", "LongParameterList", "LongMethod")

package app.pantopus.android.ui.screens.shared.content_detail.headers

import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.offset
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.heading
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import app.pantopus.android.ui.components.AvatarWithIdentityRing
import app.pantopus.android.ui.components.IdentityPillar
import app.pantopus.android.ui.components.StatusChip
import app.pantopus.android.ui.components.StatusChipVariant
import app.pantopus.android.ui.components.VerifiedBadge
import app.pantopus.android.ui.theme.PantopusColors
import app.pantopus.android.ui.theme.PantopusIcon
import app.pantopus.android.ui.theme.Spacing

/** Verification state for one identity pillar. */
enum class IdentityPillarVerificationState { Verified, Unverified }

/**
 * Identity-pillar + state pair, used by [ProfileHeader] to render the
 * Personal / Home / Business chip row.
 */
data class IdentityPillarBadge(
    val pillar: IdentityPillar,
    val state: IdentityPillarVerificationState,
) {
    val label: String
        get() =
            when (pillar) {
                IdentityPillar.Personal -> "Personal"
                IdentityPillar.Home -> "Home"
                IdentityPillar.Business -> "Business"
            }

    val chipVariant: StatusChipVariant
        get() =
            when {
                state == IdentityPillarVerificationState.Unverified -> StatusChipVariant.Neutral
                pillar == IdentityPillar.Personal -> StatusChipVariant.Personal
                pillar == IdentityPillar.Home -> StatusChipVariant.Home
                else -> StatusChipVariant.Business
            }

    val leadingIcon: PantopusIcon
        get() = if (state == IdentityPillarVerificationState.Verified) PantopusIcon.Check else PantopusIcon.Circle
}

/**
 * Centered profile header: 72dp avatar + 28dp verified badge overlay,
 * name, handle/locality row, identity-pillar chip row.
 */
@Composable
fun ProfileHeader(
    displayName: String,
    handle: String?,
    locality: String?,
    avatarUrl: String?,
    isVerified: Boolean,
    identityBadges: List<IdentityPillarBadge>,
    onBadgeTap: (IdentityPillar) -> Unit = {},
) {
    Column(
        modifier =
            Modifier
                .fillMaxWidth()
                .padding(horizontal = Spacing.s4, vertical = Spacing.s5),
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.spacedBy(Spacing.s3),
    ) {
        Box(modifier = Modifier.size(80.dp), contentAlignment = Alignment.TopStart) {
            AvatarWithIdentityRing(
                name = displayName,
                imageUrl = avatarUrl,
                identity = IdentityPillar.Personal,
                ringProgress = 1f,
                size = 72.dp,
            )
            if (isVerified) {
                Box(
                    modifier =
                        Modifier
                            .size(32.dp)
                            .offset(x = 48.dp, y = 48.dp),
                ) {
                    VerifiedBadge(size = 28.dp)
                }
            }
        }
        Text(
            text = displayName,
            fontSize = 20.sp,
            fontWeight = FontWeight.Bold,
            color = PantopusColors.appText,
            textAlign = TextAlign.Center,
            maxLines = 2,
            modifier = Modifier.semantics { heading() },
        )
        val handleAndLocality = buildHandleAndLocality(handle, locality)
        if (handleAndLocality.isNotEmpty()) {
            Text(
                text = handleAndLocality,
                fontSize = 12.sp,
                color = PantopusColors.appTextSecondary,
                maxLines = 1,
            )
        }
        if (identityBadges.isNotEmpty()) {
            Row(horizontalArrangement = Arrangement.spacedBy(Spacing.s2)) {
                identityBadges.forEach { badge ->
                    Box(
                        modifier =
                            Modifier
                                .clickable {
                                    // TODO(routing): identity-detail screens not designed yet
                                    onBadgeTap(badge.pillar)
                                }
                                .semantics {
                                    contentDescription = "${badge.label} identity, " +
                                        if (badge.state == IdentityPillarVerificationState.Verified) {
                                            "verified"
                                        } else {
                                            "not verified"
                                        }
                                },
                    ) {
                        StatusChip(text = badge.label, variant = badge.chipVariant, icon = badge.leadingIcon)
                    }
                }
            }
        }
    }
}

private fun buildHandleAndLocality(
    handle: String?,
    locality: String?,
): String =
    when {
        !handle.isNullOrEmpty() && !locality.isNullOrEmpty() -> "@$handle · $locality"
        !handle.isNullOrEmpty() -> "@$handle"
        !locality.isNullOrEmpty() -> locality
        else -> ""
    }

@Preview(showBackground = true, widthDp = 360, heightDp = 280)
@Composable
private fun ProfileHeaderPreview() {
    ProfileHeader(
        displayName = "Alex Rivera",
        handle = "alex",
        locality = "Cambridge, MA",
        avatarUrl = null,
        isVerified = true,
        identityBadges =
            listOf(
                IdentityPillarBadge(IdentityPillar.Personal, IdentityPillarVerificationState.Verified),
                IdentityPillarBadge(IdentityPillar.Home, IdentityPillarVerificationState.Verified),
                IdentityPillarBadge(IdentityPillar.Business, IdentityPillarVerificationState.Unverified),
            ),
    )
}

File: frontend/apps/ios/Pantopus/Features/Shared/ContentDetail/Headers/ProfileHeader.swift
//
//  ProfileHeader.swift
//  Pantopus
//
//  `profile_header` slot for the Public profile detail. Centered 72pt
//  avatar with 28pt VerifiedBadge overlay, name + handle/locality row,
//  identity-pillar chip row.
//

import SwiftUI

/// Verification state for one of the three identity pillars.
public enum IdentityPillarVerificationState: Sendable, Equatable {
    case verified
    case unverified
}

/// Lightweight tuple — pillar + its verification state.
public struct IdentityPillarBadge: Sendable, Identifiable, Hashable {
    public let pillar: IdentityPillar
    public let state: IdentityPillarVerificationState

    public init(pillar: IdentityPillar, state: IdentityPillarVerificationState) {
        self.pillar = pillar
        self.state = state
    }

    public var id: String {
        switch pillar {
        case .personal: "personal"
        case .home: "home"
        case .business: "business"
        }
    }

    var label: String {
        switch pillar {
        case .personal: "Personal"
        case .home: "Home"
        case .business: "Business"
        }
    }

    var chipVariant: StatusChipVariant {
        switch (pillar, state) {
        case (.personal, .verified): .personal
        case (.home, .verified): .home
        case (.business, .verified): .business
        case (_, .unverified): .neutral
        }
    }

    var leadingIcon: PantopusIcon {
        state == .verified ? .check : .circle
    }
}

/// Centered profile header: 72pt avatar + verified-badge overlay, name,
/// handle/locality, identity-pillar chip row.
@MainActor
public struct ProfileHeader: View {
    private let displayName: String
    private let handle: String?
    private let locality: String?
    private let avatarURL: URL?
    private let isVerified: Bool
    private let identityBadges: [IdentityPillarBadge]
    /// `nil` renders the badges as non-tappable status chips. Pass a
    /// real handler to make them open an identity-detail surface.
    private let onBadgeTap: (@MainActor (IdentityPillar) -> Void)?

    public init(
        displayName: String,
        handle: String?,
        locality: String?,
        avatarURL: URL?,
        isVerified: Bool,
        identityBadges: [IdentityPillarBadge],
        onBadgeTap: (@MainActor (IdentityPillar) -> Void)? = nil
    ) {
        self.displayName = displayName
        self.handle = handle
        self.locality = locality
        self.avatarURL = avatarURL
        self.isVerified = isVerified
        self.identityBadges = identityBadges
        self.onBadgeTap = onBadgeTap
    }

    public var body: some View {
        VStack(spacing: Spacing.s3) {
            ZStack(alignment: .bottomTrailing) {
                AvatarWithIdentityRing(
                    name: displayName,
                    imageURL: avatarURL,
                    identity: .personal,
                    ringProgress: 1,
                    size: 72
                )
                if isVerified {
                    VerifiedBadge(size: 28).offset(x: 4, y: 4)
                }
            }
            .frame(width: 80, height: 80)

            Text(displayName)
                .font(.system(size: PantopusTextStyle.h3.size, weight: .bold))
                .tracking(-0.25)
                .foregroundStyle(Theme.Color.appText)
                .lineLimit(2)
                .multilineTextAlignment(.center)
                .accessibilityAddTraits(.isHeader)

            if handle != nil || locality != nil {
                Text(handleAndLocality)
                    .font(.system(size: PantopusTextStyle.caption.size, weight: .regular))
                    .foregroundStyle(Theme.Color.appTextSecondary)
                    .lineLimit(1)
            }

            if !identityBadges.isEmpty {
                HStack(spacing: Spacing.s2) {
                    ForEach(identityBadges) { badge in
                        if let onBadgeTap {
                            Button { onBadgeTap(badge.pillar) } label: {
                                StatusChip(
                                    badge.label,
                                    variant: badge.chipVariant,
                                    icon: badge.leadingIcon
                                )
                            }
                            .buttonStyle(.plain)
                            .accessibilityLabel(
                                "\(badge.label) identity, \(badge.state == .verified ? "verified" : "not verified")"
                            )
                        } else {
                            StatusChip(
                                badge.label,
                                variant: badge.chipVariant,
                                icon: badge.leadingIcon
                            )
                            .accessibilityElement()
                            .accessibilityLabel(
                                "\(badge.label) identity, \(badge.state == .verified ? "verified" : "not verified")"
                            )
                        }
                    }
                }
            }
        }
        .frame(maxWidth: .infinity)
        .padding(.horizontal, Spacing.s4)
        .padding(.top, Spacing.s4)
        .padding(.bottom, Spacing.s5)
        .accessibilityElement(children: .contain)
    }

    private var handleAndLocality: String {
        switch (handle, locality) {
        case let (h?, l?) where !h.isEmpty && !l.isEmpty: "@\(h) · \(l)"
        case let (h?, _) where !h.isEmpty: "@\(h)"
        case let (_, l?) where !l.isEmpty: l
        default: ""
        }
    }
}

#Preview("Verified + all badges") {
    ProfileHeader(
        displayName: "Alex Rivera",
        handle: "alex",
        locality: "Cambridge, MA",
        avatarURL: nil,
        isVerified: true,
        identityBadges: [
            IdentityPillarBadge(pillar: .personal, state: .verified),
            IdentityPillarBadge(pillar: .home, state: .verified),
            IdentityPillarBadge(pillar: .business, state: .unverified)
        ]
    )
    .padding(.vertical, Spacing.s4)
    .background(Theme.Color.appBg)
}
